# 161476173 login user

wiztemple opened the pull request:
### What does this PR do?
Logs in a registered user to the application

### Description of Tasks to be completed
- create a method to login user
- generate token for user

### How should this be manually tested?
- git clone the repository
- cd into the the repo folder
- run git checkout ft-users-receive-jwt-successful-161476173

Open up postman and hit the following localhost:5000/api/v1/users/signin route with the json input
```
{
   email: "wisdom@gmail.com",
   password: "weirdcode",
}
```

Run npm test

### What are the relevant Pivotal Tracker Stories?
- [#161476173](https://www.pivotaltracker.com/story/show/161476173)

### Screenshots(if appropriate)
<img width="933" alt="screen shot 2018-10-30 at 10 31 51 am" src="https://user-images.githubusercontent.com/34472497/47708853-48ef0e00-dc2f-11e8-96f8-1ab864e98958.png">

<img width="1369" alt="screenshot 2018-10-29 at 7 07 08 pm" src="https://user-images.githubusercontent.com/25081005/47672490-75187980-dbb2-11e8-8ef1-e357c4884c9a.png">

<img width="824" alt="screen shot 2018-10-30 at 7 24 11 am" src="https://user-images.githubusercontent.com/34472497/47700008-1df7c080-dc15-11e8-92d2-f6f210c4f4ea.png">

PriscillaSam commented:
This looks alright.

TheoOkafor commented on server/controllers/user.js:
```diff
@@ -49,6 +49,42 @@ class User {
+    return badHttpResponse(
+      response,
+      404,
+      'User does not exist'
```
I think the error message should also inform the user about the actions to take. Consider changing this to ```"This account does not exist. Consider sign up."```

TheoOkafor commented on server/tests/controllerTests/user.test.js:
```diff
@@ -36,3 +37,62 @@ describe('POST api/users/signup', () => {
+  it('should not sign in user without password', async () => {
+    const response = await chai.request(app)
+      .post(signinRoute)
+      .send({
```
Consider storing this in the mockData.js file created by Priscilla

TheoOkafor commented on server/tests/controllerTests/user.test.js:
```diff
@@ -36,3 +37,62 @@ describe('POST api/users/signup', () => {
+  it('should not sign in user with an incorrect password', async () => {
+    const response = await chai.request(app)
+      .post(signinRoute)
+      .send({
```
This one too and the other test data.

TheoOkafor commented on server/controllers/user.js:
```diff
@@ -49,6 +49,42 @@ class User {
+      return goodHttpResponse(
+        response,
+        200,
+        `Hello ${findUser.username}, Welcome Back!`,
```
Nice message Gents. I like it 👍

PriscillaSam commented on server/controllers/user.js:
```diff
@@ -49,6 +49,42 @@ class User {
+        response,
+        200,
+        `Hello ${findUser.username}, Welcome Back!`,
+        token
```
@wiztemple @ucheg6 please put the token in an object like this `{ token }` so that the data object from the response has the attribute as `token` and not `output`.

seyi-adeleke commented on server/controllers/user.js:
```diff
@@ -49,6 +49,42 @@ class User {
+      if (validatePassword === false) {
+        return badHttpResponse(
+          response,
+          403,
```
This should be a 401

seyi-adeleke commented on server/middlewares/validations.js:
```diff
@@ -17,4 +17,14 @@ const inputValidator = (request, response, next) => {
+    'email',
+    'password',
+  ]);
+  if (isValid === false) {
```
`!isValid` for brevity

seyi-adeleke commented on server/routes/users.js:
```diff
@@ -1,9 +1,10 @@
 
 const router = new Router();
 
 router.post('/users/signup', inputValidator, User.signup);
```
`inputValidator` should be `signupValidator` then

ucheg6 commented:
@seyi-adeleke all your feedback has been addressed

seyi-adeleke commented on package.json:
```diff
@@ -42,6 +42,7 @@
     "babel-plugin-add-module-exports": "^1.0.0",
     "babel-plugin-istanbul": "^5.1.0",
     "babel-plugin-transform-runtime": "^6.23.0",
+    "babel-polyfill": "^6.26.0",
```
Why did you add this package?

ucheg6 replied:
I checked in with Priscilla and Theo and they said they put it there to enable the async/await syntax

ucheg6 replied:
They also said that the package is not needed again because of the presence of babel-plugin-transform-runtime in .babelrc. So we agreed I should remove it.

ucheg6 committed:
ft(signin):login user

- create a method to login user
- generate token for user
- validate login
- write tests for login route
